pkg/dependenciesdistributor/dependencies_distributor.go

```diff
@@ -0,0 +1,551 @@
+}
+
+// OnUpdate handles object update event and push the object to queue.
+func (d *DependenciesDistributor) OnUpdate(oldObj, newObj interface{}) {
+	d.OnAdd(newObj)
+}
+
+// OnDelete handles object delete event and push the object to queue.
+func (d *DependenciesDistributor) OnDelete(obj interface{}) {
+	d.OnAdd(obj)
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the key.
+// The key will be re-queued if an error is non-nil.
+func (d *DependenciesDistributor) Reconcile(key util.QueueKey) error {
+	clusterWideKey, ok := key.(keys.ClusterWideKey)
+	if !ok {
+		klog.Error("invalid key")
+		return fmt.Errorf("invalid key")
+	}
+	klog.Infof("DependenciesDistributor start to reconcile object: %s", clusterWideKey)
+
+	bindingObjectList, err := d.resourceBindingLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	bindingList, err := convertObjectsToResourceBindings(bindingObjectList)
+	if err != nil {
+		return err
+	}
+
+	var errs []error
+	for _, binding := range bindingList {
+		if !binding.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		matched, err := dependentObjectReferenceMatches(clusterWideKey, binding)
+		if err != nil {
+			klog.Errorf("failed to evaluate if binding(%s/%s) need to sync dependencies: %v", binding.Namespace, binding.Name, err)
+			errs = append(errs, err)
+			continue
+		} else if !matched {
+			klog.V(4).Infof("no need to sync binding(%s/%s)", binding.Namespace, binding.Name)
+			continue
+		}
+
+		bindingKey, err := detector.ClusterWideKeyFunc(binding)
+		if err != nil {
+			klog.Errorf("failed to generate cluster wide key for binding %s/%s: %v", binding.Namespace, binding.Name, err)
+			errs = append(errs, err)
+			continue
+		}
+
+		d.bindingReconcileWorker.Add(bindingKey)
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+// dependentObjectReferenceMatches tells if the given object is referred by current resource binding.
+func dependentObjectReferenceMatches(objectKey keys.ClusterWideKey, referenceBinding *workv1alpha2.ResourceBinding) (bool, error) {
+	dependencies, exist := referenceBinding.Annotations[bindingDependenciesAnnotationKey]
+	if !exist {
+		return false, nil
+	}
+
+	var dependenciesSlice []configv1alpha1.DependentObjectReference
+	err := json.Unmarshal([]byte(dependencies), &dependenciesSlice)
+	if err != nil {
+		return false, err
+	}
+
+	if len(dependenciesSlice) == 0 {
+		return false, nil
+	}
+
+	for _, dependence := range dependenciesSlice {
+		if objectKey.Version == dependence.APIVersion &&
+			objectKey.Kind == dependence.Kind &&
+			objectKey.Namespace == dependence.Namespace &&
+			objectKey.Name == dependence.Name {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// OnResourceBindingAdd handles object add event.
+func (d *DependenciesDistributor) OnResourceBindingAdd(obj interface{}) {
+	key, err := detector.ClusterWideKeyFunc(obj)
+	if err != nil {
+		return
+	}
+
+	d.bindingReconcileWorker.Add(key)
+}
```

Comment:
Perhaps we can narrow down the binding that adds the queue. Consider that only when `binding.Spec.PropagateDeps` is true.

For the update scenario, the `binding.Spec.PropagateDeps` field of at least one object must be set to true.